include/ofi_signal.h

```diff
@@ -87,18 +104,72 @@ static inline void fd_signal_free(struct fd_signal *signal)
 static inline void fd_signal_set(struct fd_signal *signal)
 {
 	char c = 0;
-	if (signal->wcnt == signal->rcnt) {
-		if (ofi_write_socket(signal->fd[FI_WRITE_FD], &c, sizeof c) == sizeof c)
-			signal->wcnt++;
+	bool cas = ofi_atomic_cas_bool_strong32(&signal->flag,
+					     FI_SIGNAL_UNSET,
+					     FI_SIGNAL_WRITE_PREPARE);
+	if (cas) {
+		int wr_rc = ofi_write_socket(signal->fd[FI_WRITE_FD], &c,
+					     sizeof c);
+		if (wr_rc == sizeof c) {
+#if ENABLE_DEBUG
+			int cnt = ofi_atomic_inc32(&signal->debug_cnt);
+			assert(cnt == 1);
+#endif
+			cas = ofi_atomic_cas_bool_strong32(&signal->flag,
+							   FI_SIGNAL_WRITE_PREPARE,
+							   FI_SIGNAL_SET);
+			assert(cas == true);
+		}
+		else {
+			/* XXX setting the signal failed. A polling thread
+			 * will not be woken up now. And the system might
+			 * get stuck.
+			 * Also, typically this will be totally
+			 * untested code path, as it basically will never
+			 * come up.
+			 */
+			cas = ofi_atomic_cas_bool_strong32(&signal->flag,
+							   FI_SIGNAL_WRITE_PREPARE,
+							   FI_SIGNAL_UNSET);
+			assert(cas == true);
+		}
 	}
 }
 
 static inline void fd_signal_reset(struct fd_signal *signal)
 {
 	char c;
-	if (signal->rcnt != signal->wcnt) {
-		if (ofi_read_socket(signal->fd[FI_READ_FD], &c, sizeof c) == sizeof c)
-			signal->rcnt++;
+	while (true) {
+		int flag = ofi_atomic_get32(&signal->flag);
+
+		if (flag == FI_SIGNAL_UNSET || flag > FI_SIGNAL_SET)
+			break;
+
+		/* XXX: Switch to _weak as this is in a while loop */
+		bool cas = ofi_atomic_cas_bool_strong32(&signal->flag,
+							FI_SIGNAL_SET,
+							FI_SIGNAL_READ_PREPARE);
+		if (cas) {
+			int read_rc = ofi_read_socket(signal->fd[FI_READ_FD], &c,
+						      sizeof c);
+			if (read_rc == sizeof c) {
+#if ENABLE_DEBUG
+				int cnt = ofi_atomic_dec32(&signal->debug_cnt);
+				assert(cnt == 0);
+#endif
+				cas = ofi_atomic_cas_bool_strong32(&signal->flag,
+							FI_SIGNAL_READ_PREPARE,
+							FI_SIGNAL_UNSET);
+				assert(cas == true);
+				break;
+			} else {
+				/* Could this ever happen with an alive socketd? */
+				cas = ofi_atomic_cas_bool_strong32(&signal->flag,
+						FI_SIGNAL_READ_PREPARE,
+						FI_SIGNAL_SET);
```

Comment:
atomic set.  This path should basically never occur in practice.  But if it does, we should probably abort the loop here to avoid spinning forever through this error path.